test/integration/finance/api/invoices_test.rb

```diff
@@ -1,210 +1,199 @@
+      buyer = @buyer || FactoryBot.create(:buyer_account, provider_account: @provider)
 
-        @context = ''
-        host! @provider.admin_domain
+      25.times do
+        FactoryBot.create(:invoice, provider_account: @provider, buyer_account: buyer)
       end
 
-      @@common_cases.call
+      get "/api/#{@context}invoices.xml?provider_key=#{@key}&page=1&per_page=21"
+
+      assert_response :success
+      assert_select 'invoices > pagination[total_pages=?]', '2'
+      assert_select 'invoices > pagination[current_page=?]', '1'
+      assert_select 'invoices > pagination[per_page=?]', '21'
+      assert_select 'invoices > invoice', count: 21
+    end
+
+    test 'redirect when asked for PDF' do
+      setup_existing_invoices_test
+      get "/api/#{@context}invoices/#{@invoice.id}.pdf?provider_key=#{@key}"
+      assert_response :found
     end
 
-    context 'Invoice API[by buyers]' do
-      setup do
-        @provider = FactoryBot.create(:provider_account)
-        @buyer = FactoryBot.create(:buyer_account, :provider_account => @provider)
+    test 'filter by month' do
+      setup_existing_invoices_test
+      setup_filter_tests
+      @invoice.update(period: Month.new(1022,5))
+
+      get "/api/#{@context}invoices.xml?provider_key=#{@key}&month=1022-05"
 
-        @provider.create_billing_strategy
-        @provider.save!
-        @key = @provider.api_key
+      assert_response :success
 
-        @context = "accounts/#{@buyer.id}/"
-        host! @provider.admin_domain
+      assert_select 'invoices > invoice', count: 1
+    end
+
+    test 'filter by state' do
+      setup_existing_invoices_test
+      setup_filter_tests
+      @invoice.cancel!
+
+      get "/api/#{@context}invoices.xml?provider_key=#{@key}&state=cancelled"
+
+      assert_response :success
+
+      assert_select 'invoices > invoice', count: 1
+      assert_select 'invoices > invoice > id', text: @invoice.id.to_s
+      assert_select 'invoices > invoice > state', text: 'cancelled'
+    end
+
+    def setup_existing_invoices_test
+      @buyer ||= FactoryBot.create(:buyer_account, provider_account: @provider)
+      @invoice = FactoryBot.create(:invoice, provider_account: @provider, buyer_account: @buyer)
+      2.times { FactoryBot.create(:line_item_plan_cost, invoice: @invoice, name: 'fake', cost: 10.0) }
+    end
+
+    def setup_filter_tests
+      buyer = @buyer || FactoryBot.create(:buyer_account, provider_account: @provider)
+
+      2.times do
+        FactoryBot.create(:invoice, provider_account: @provider, buyer_account: buyer)
       end
+    end
+  end
+
+  class InvoicesNoScopeTest < InvoicesTestCommonCases
+    def setup
+      super
+      @context = ''
+    end
+  end
 
-      @@common_cases.call
+  class InvoicesBuyersScopeTest < InvoicesTestCommonCases
+    def setup
+      super
+      @context = "accounts/#{@buyer.id}/"
     end
+  end
+
+  def self.runnable_methods
+    return [] if self == InvoicesTestCommonCases
 
+    super
```

Comment:
Actually is it possible to rename `InvoicesTestCommonCases` to `InvoicesUnscopedTest`, then `InvoicesBuyersScopeTest` inherit from `InvoicesUnscopedTest`. In this way it will not be needed to deal with implementation details like `#runnable_methods`. WDYT?

Reply:
Absolutely, good catch!